### ISSUE TITLE
A better solidity implementation for the conviction formula

### ISSUE THREAD

**Issue by sembrestels** [1Hive/conviction-voting-app#21]
The conviction formula that @dapplion came up on [ETHParis](https://hackmd.io/jWlia12DQ6SBLpKGLP9_-g) is as follows:

<p align="center"><img src="https://latex.codecogs.com/gif.latex?y_t=\alpha^ty_0+x\frac{1-\alpha^t}{1-\alpha}" /></p>

This has been very useful to derivate other functions such as the max amount of conviction a proposal can have, the time it is required for a proposal to pass with current conviction and stakes, etc.

He also provided a Solidity implementation that deals with the lack of decimals in that language. He uses a parameter `D = 10` to augment <img src="https://latex.codecogs.com/gif.latex?\alpha" /> and <img src="https://latex.codecogs.com/gif.latex?\alpha^t" />, so instead of <img src="https://latex.codecogs.com/gif.latex?\alpha=0.9" /> we have <img src="https://latex.codecogs.com/gif.latex?\alpha_D=9" />:

<p align="center"><img src="https://latex.codecogs.com/gif.latex?y_t=\frac{1}{D^t}[\alpha_D^t%20y_0+xD\frac{D^t-\alpha_D^t}{D-\alpha_D}]"/></p>

But to do so we need to calculate <img src="https://latex.codecogs.com/gif.latex?D^t"/> and <img src="https://latex.codecogs.com/gif.latex?\alpha%20D^t"/>. In Solidity:
```go
// Predefined params
pragma solidity >=0.4.22 <0.6.0;
contract ConvictionComputation {
    uint D = 10;
    uint aD = 9;
    // Sample state
    uint y0 = 0;
    uint x = 1e18;
    function conviction(uint t) public view returns (uint) {
        uint Dt = D**t;
        uint aDt = aD**t;
        uint y = (aDt * y0 + (x*D*(Dt-aDt))/(D-aD)) / Dt;
        return y;
    }
}
```
The problem with that approach is that uint variables `Dt` and `aDt` overflow when `t > 77` and `t > 90` respectively. In other words, after 77 units of time from last stake or withdraw, the function becomes unreliable because it tries to calculate a number beyond the maximum unsigned integer (2^256-1).

A single approach to deal with that is to consider `t=77` the last secure number, and consider <img src="https://latex.codecogs.com/gif.latex?\alpha^t=0" /> from that moment onwards, so the conviction formula for `t > 77` is:

<p align="center"><img src="https://latex.codecogs.com/gif.latex?y_t=x\frac{1}{1-\alpha}" /></p>

This is the approach we take right now in our code:
https://github.com/1Hive/conviction-voting-app/blob/7c6a13006c29afb6dc35f2b22b1ec022299ebdcc/contracts/ConvictionVotingApp.sol#L232-L237

It is a good enough solution for now, but we should find a better solution because we are losing a lot of precision if we consider <img src="https://latex.codecogs.com/gif.latex?\alpha^t" /> to be 0 when it really is <img src="https://latex.codecogs.com/gif.latex?\alpha^t=0.9^{78}=0.000269722" />. If we don't find anything better, we can just consider an implementation quirk, in which tokens are fully activated after 77 units of time.

**Comment by sembrestels** [comment]
After writing this I changed my mind. I think an even simpler solution is to use `t=77` when `t>77`, so we cap conviction on what it reaches during its first 77 units of time from the last stake or withdrawal.

On the other hand, we are still looking for a better solidity function to avoid all that problem.

**Comment by sembrestels** [comment]
In 68a49ed I implement an alternative implementation to ETHParis' solidity implementation for conviction formula, which sets a constant scaling factor of <img src="https://latex.codecogs.com/gif.latex?2^{128}" /> instead of an exponential <img src="https://latex.codecogs.com/gif.latex?D^t" />:

<p align="center"><img src="https://latex.codecogs.com/gif.latex?y=\frac{1}{2^{128}}\left(2^{128}\alpha^ty_0+xD\frac{2^{128}-2^{128}\alpha^t}{D-\alpha%20D}+2^{127}\right)" /></p>

The `+2^127` at the end is used for rounding. The tests I've written show that the formula maintains the relative error below the 0.1% for a wide variety of parameters.

**Comment by sembrestels** [comment]
Fixed by #21.
